dhis2/dhis2-core
# feat: invalidate user session when user's role memberships changes

# Summary 
* The user will now be logged out of any current sessions when their role memberships are changed by an admin via the user admin UI or via an API call. Or if any of the assigned roles changes authorities, similarly. Without this feature, the user's new role assignments or role changes will not be effective in the ongoing sessions.

# Automatic test
* UserControllerTest#updateRolesShouldInvalidateUserSessions()

# Manual test
### Test role assignment will log out the users
1. Create two users, one admin and one non-admin.
2. As the admin user, create a superuser role that ass the “ALL” authority.
3. As the admin user, assign the superuser role to the non-admin user.
4. Log in as the non-admin user and try to edit a user in the user admin, this should work.
5. As admin (in another browser), remove the superuser role from the non-admin user.
6. As the non-admin user, try to go to any other page, you should now observe that you will be logged out and redirected to the login page.

### Test role changes will log out the users
1. Create two users, one admin and one non-admin.
2. As the admin user, create a superuser role that ass the “ALL” authority.
3. As the admin user, assign the superuser role to the non-admin user.
4. Log in as the non-admin user and try to edit a user in the user admin, this should work.
5. As admin (in another browser), remove the superuser “ALL” authority from the superuser role.
7. As the non-admin user, try to go to any other page, you should now observe that you will be logged out and 

### Jira:
[DHIS2-10334](https://dhis2.atlassian.net/browse/DHIS2-10334)

[DHIS2-10334]: https://dhis2.atlassian.net/browse/DHIS2-10334?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
